sdb/commands/spl/internal/kmem_helpers.py

```diff
@@ -116,20 +121,88 @@ def entry_size(cache: drgn.Object) -> int:
-    assert cache.type_.type_name() == 'spl_kmem_cache_t *'
+    assert sdb.type_canonical_name(cache.type_) == 'struct spl_kmem_cache *'
     total_mem = total_memory(cache)
     if total_mem == 0:
         return 0
     return int((active_memory(cache) / total_mem) * 100)
+
+
+def sko_from_obj(cache: drgn.Object, obj: drgn.Object) -> drgn.Object:
+    assert sdb.type_canonical_name(cache.type_) == 'struct spl_kmem_cache *'
+    cache_obj_align = cache.skc_obj_align.value_()
+    return sdb.create_object(
+        'spl_kmem_obj_t *',
+        obj.value_() +
+        p2.p2roundup_typed(object_size(cache), cache_obj_align, 'uint32_t'))
+
+
+def spl_aligned_obj_size(cache: drgn.Object) -> int:
+    assert sdb.type_canonical_name(cache.type_) == 'struct spl_kmem_cache *'
+    cache_obj_align = cache.skc_obj_align.value_()
+    spl_obj_type_size = sdb.type_canonicalize_size('spl_kmem_obj_t')
+    return p2.p2roundup_typed(
+        object_size(cache), cache_obj_align, 'uint32_t') + p2.p2roundup_typed(
+            spl_obj_type_size, cache_obj_align, 'uint32_t')
+
+
+def spl_aligned_slab_size(cache: drgn.Object) -> int:
+    assert sdb.type_canonical_name(cache.type_) == 'struct spl_kmem_cache *'
+    cache_obj_align = cache.skc_obj_align.value_()
+    spl_slab_type_size = sdb.type_canonicalize_size('spl_kmem_slab_t')
+    return p2.p2roundup_typed(spl_slab_type_size, cache_obj_align, 'uint32_t')
+
+
+def for_each_offslab_object_in_slab(slab: drgn.Object) -> Iterable[drgn.Object]:
+    assert sdb.type_canonical_name(slab.type_) == 'struct spl_kmem_slab *'
+    #
+    # ZFSonLinux initially implemented these caches for certain cases that
+    # never showed up, as a result we leave this code path unimplemented
+    # for now.
+    #
+    raise NotImplementedError
+
+
+def for_each_onslab_object_in_slab(slab: drgn.Object) -> Iterable[drgn.Object]:
+    assert sdb.type_canonical_name(slab.type_) == 'struct spl_kmem_slab *'
+    cache = slab.sks_cache
+    sks_size = spl_aligned_slab_size(cache)
+    spl_obj_size = spl_aligned_obj_size(cache)
+
+    for i in range(slab.sks_objs.value_()):
+        obj = sdb.create_object('void *',
+                                slab.value_() + sks_size + (i * spl_obj_size))
+        #
+        # If the sko_list of the object is empty, it means that
+        # this object is not part of the slab's internal free list
+        # and therefore it is allocated.
+        #
+        sko = sko_from_obj(cache, obj)
+        assert sko.sko_magic.value_() == 0x20202020  # SKO_MAGIC
+        if spl_util.is_list_empty(sko.sko_list):
+            yield obj
+
+
+def for_each_object_in_spl_cache(cache: drgn.Object) -> Iterable[drgn.Object]:
+    assert sdb.type_canonical_name(cache.type_) == 'struct spl_kmem_cache *'
+
+    slab_object_iterator = for_each_onslab_object_in_slab
+    if 'KMC_OFFSLAB' in list(for_each_slab_flag_in_cache(cache)):
+        slab_object_iterator = for_each_offslab_object_in_slab
```

Comment:
Why bother with  `for_each_offslab_object_in_slab`, instead of having a check here that KMC_OFFSLAB is not set?  That way you wouldn't need `slab_object_iterator`.